android/src/main/java/com/squareup/sdk/reader/react/StoreCustomerCardModule.java

```diff
@@ -0,0 +1,104 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package com.squareup.sdk.reader.react;
+
+import android.app.Activity;
+import android.os.Handler;
+import android.os.Looper;
+import com.facebook.react.bridge.Promise;
+import com.facebook.react.bridge.ReactApplicationContext;
+import com.facebook.react.bridge.ReactContextBaseJavaModule;
+import com.facebook.react.bridge.ReactMethod;
+import com.squareup.sdk.reader.ReaderSdk;
+import com.squareup.sdk.reader.checkout.Card;
+import com.squareup.sdk.reader.core.CallbackReference;
+import com.squareup.sdk.reader.core.Result;
+import com.squareup.sdk.reader.core.ResultError;
+import com.squareup.sdk.reader.crm.StoreCardActivityCallback;
+import com.squareup.sdk.reader.crm.StoreCustomerCardErrorCode;
+import com.squareup.sdk.reader.react.internal.ErrorHandlerUtils;
+import com.squareup.sdk.reader.react.internal.ReaderSdkException;
+import com.squareup.sdk.reader.react.internal.converter.CardConverter;
+
+class StoreCustomerCardModule extends ReactContextBaseJavaModule {
+    // Define all the store customer card debug codes and messages below
+    // These error codes and messages **MUST** align with iOS error codes and javascript error codes
+    // Search KEEP_IN_SYNC_STORE_CUSTOMER_CARD_ERROR to update all places
+
+    // react native module debug error codes
+    private static final String RN_STORE_CUSTOMER_CARD_ALREADY_IN_PROGRESS = "rn_store_customer_card_already_in_progress";
+
+    // react native module debug messages
+    private static final String RN_MESSAGE_STORE_CUSTOMER_CARD_ALREADY_IN_PROGRESS = "A store customer card operation is already in progress. Ensure that the in-progress store customer card is completed before calling startStoreCardAsync again.";
+
+    private volatile CallbackReference storeCardCallbackRef;
+    private final Handler mainLooperHandler;
+
+    public StoreCustomerCardModule(ReactApplicationContext reactContext) {
+        super(reactContext);
+        mainLooperHandler = new Handler(Looper.getMainLooper());
+    }
+
+    @Override
+    public String getName() {
+        return "RNReaderSDKStoreCustomerCard";
+    }
+
+    @ReactMethod
+    public void startStoreCard(final String customerId, final Promise promise) {
+        if (storeCardCallbackRef != null) {
+            String errorJsonMessage = ErrorHandlerUtils.createNativeModuleError(RN_STORE_CUSTOMER_CARD_ALREADY_IN_PROGRESS, RN_MESSAGE_STORE_CUSTOMER_CARD_ALREADY_IN_PROGRESS);
+            promise.reject(ErrorHandlerUtils.USAGE_ERROR, new ReaderSdkException(errorJsonMessage));
+            return;
+        }
+
+        StoreCardActivityCallback storeCardActivityCallback = new StoreCardActivityCallback() {
```

Comment:
It might not matter much, and the need to access the promise might force you towards an "extra" allocation anyway... but you could avoid doing the allocation/GC cycle on every callback object by allocating one callback the first time we're called (or on create, or something), and instead of GC'ing and reallocating, store the promise in a field (like you're doing now with storeCardCallbackRef) and just pull that in the one, long-lived callback object.

Reply:
I remember I discussed this with someone before for other modules which take similar approach. We decided to do so to avoid handling the state of promise given we assume this allocation should be super cheap.